### PR TITLE
Implement Write

### PR DESCRIPTION
**Pull request by machariamuguku**
- closes https://github.com/onaio/canopy/issues/2944
- Implement airbyte write spec
- Tests

### CODE REVIEW

**Review by ukanga — CHANGES_REQUESTED** [review]

File: destination_dhis2/client.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    def flush(self) -> None:
+        # TODO: Handle retry?
+        # best place to handle retry imo
+        if len(self.write_buffer) > 0:
+            try:
+                response = self._batch_write(self.write_buffer)
+                response.raise_for_status()
+                self.write_buffer.clear()
+            except Exception as e:
+                raise e
```
[the remark this answers is not in the document]
Comment (machariamuguku) [reply destination_dhis2/client.py:122]: The errors are supposed to be caught downstream. I'll remove the try-except block

File: destination_dhis2/destination.py [hunk 2]
```diff
@@ -28,43 +29,82 @@ def write(
...
-        :return: Iterable of AirbyteStateMessages wrapped in AirbyteMessage structs
-        """
+                    if client.buffer_is_full():
+                        try:
+                            client.flush()
+                        except Exception as e:
+                            airbyteLogger.error(
+                                f"Exception flushing AirbyteRecordMessage: {e}"
+                            )
+                            raise e
```
[the remark this answers is not in the document]
Comment (machariamuguku) [reply destination_dhis2/destination.py:62]: I'm planning to handle re-try for transient exceptions (Timeout etc) [here](https://github.com/onaio/canopy/issues/3050). Will catching the broad RequestException suffice? There's no special handling expected for granular request exceptions other than logging and then re-raising them

[the remark this answers is not in the document]
Comment (ukanga) [reply destination_dhis2/destination.py:62]: Yes. That's should do.

File: destination_dhis2/client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (machariamuguku) [reply destination_dhis2/client.py]: https://github.com/onaio/destination-dhis2/pull/2/commits/b114d7c7c577c52962c28e08d45a0b04623c3d7a

[the remark this answers is not in the document]
Comment (machariamuguku) [reply destination_dhis2/client.py]: https://github.com/onaio/destination-dhis2/pull/2/commits/fcdbb40b8a40fc8e6c670f1d933f9a52b6fa51dc

File: destination_dhis2/destination.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (machariamuguku) [reply destination_dhis2/destination.py]: https://github.com/onaio/destination-dhis2/pull/2/commits/fcdbb40b8a40fc8e6c670f1d933f9a52b6fa51dc

File: destination_dhis2/destination.py [hunk 3]
```diff
@@ -81,7 +121,7 @@ def check(
                     response.raise_for_status()
             return AirbyteConnectionStatus(status=Status.SUCCEEDED)
 
-        except RequestException as req_err:
+        except (RequestException, Exception) as req_err:
```
[the remark this answers is not in the document]
Comment (machariamuguku) [reply destination_dhis2/destination.py]: https://github.com/onaio/destination-dhis2/pull/2/commits/fcdbb40b8a40fc8e6c670f1d933f9a52b6fa51dc

File: destination_dhis2/destination.py [hunk 4]
```diff
@@ -28,43 +29,82 @@ def write(
...
-        then the source is given the last state message output from this method as the starting point of the next sync.
+                    # add to buffer
+                    client.queue_write_operation(dataValue=data)
 
-        :param config: dict of JSON configuration matching the configuration declared in spec.json
-        :param configured_catalog: The Configured Catalog describing the schema of the data being received and how it should be persisted in the
-                                    destination
-        :param input_messages: The stream of input messages received from the source
-        :return: Iterable of AirbyteStateMessages wrapped in AirbyteMessage structs
-        """
```
[the remark this answers is not in the document]
Comment (machariamuguku) [reply destination_dhis2/destination.py]: [`8deb8fb` (#2)](https://github.com/onaio/destination-dhis2/pull/2/commits/8deb8fb78e779fc29a9e04c26c788ae29b17b66f)

**Pull request merged**